### PR TITLE
[SPARK-25196][SPARK-27251][SQL][FOLLOWUP] Add synchronized for InMemoryRelation.statsOfPlanToCache

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This is a follow-up of #24047; to follow the `CacheManager.cachedData` lock semantics, this pr wrapped the `statsOfPlanToCache` update with `synchronized`.

## How was this patch tested?
Pass Jenkins

### CODE REVIEW

**Comment by maropu** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala [hunk 1]
```diff
@@ -163,12 +161,15 @@ class CacheManager extends Logging {
...
     val (rowCount, newColStats) =
       CommandUtils.computeColumnStats(sparkSession, relation, column)
-    val oldStats = relation.statsOfPlanToCache
-    val newStats = oldStats.copy(
-      rowCount = Some(rowCount),
-      attributeStats = AttributeMap((oldStats.attributeStats ++ newColStats).toSeq)
-    )
-    relation.statsOfPlanToCache = newStats
+
+    relation.synchronized {
```
Comment (attilapiros) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala]: I would move this logic into the class `InMemoryRelation`. Like:

```scala
def updateStats(rowCount: Long, newColStats: Map[Attribute, ColumnStat]): Unit = synchronized {
    val newStats = statsOfPlanToCache.copy(
      rowCount = Some(rowCount),
      attributeStats = AttributeMap((statsOfPlanToCache.attributeStats ++ newColStats).toSeq)
    )
    statsOfPlanToCache = newStats
  }
```

This way both the write and the `synchronized` is in the member function of the `InMemoryRelation` (I think it is bit more readable). What is your opinion?

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala]: Looks reasonable, I'll update soon.

**Comment by dongjoon-hyun** [comment]
Retest this please

**Comment by dongjoon-hyun** [comment]
Retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala [hunk 2]
```diff
@@ -145,37 +145,60 @@ object InMemoryRelation {
...
+    relation
   }
 }
 
 case class InMemoryRelation(
     output: Seq[Attribute],
     @transient cacheBuilder: CachedRDDBuilder,
-    override val outputOrdering: Seq[SortOrder])(
-    @volatile var statsOfPlanToCache: Statistics)
+    override val outputOrdering: Seq[SortOrder])
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala]: Sorry, but could you explain why we are removing `statsOfPlanToCache` from here? IIUC, Scala-2.11 issue is just about removing `@volatile`, isn't it?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala [hunk 3]
```diff
@@ -145,37 +145,60 @@ object InMemoryRelation {
...
   extends logical.LeafNode with MultiInstanceRelation {
 
+  @volatile var statsOfPlanToCache: Statistics = null
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala]: Got it. This PR aims to hide `statsOfPlanToCache`.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @maropu and @attilapiros .
Merged to master.

**Pull request closed without merging**